scripts/zkai_d128_recursive_pcd_route_selector_gate.py

```
@@ -0,0 +1,910 @@
+    writer = csv.DictWriter(out, fieldnames=TSV_COLUMNS, delimiter="\t", lineterminator="\n")
+    writer.writeheader()
+    for route in payload["route_table"]:
+        writer.writerow({column: route[column] for column in TSV_COLUMNS})
+    return out.getvalue()
+
+
+def _ensure_repo_relative(path: pathlib.Path | None, *, field: str) -> pathlib.Path | None:
+    if path is None:
+        return None
+    if path.is_absolute():
+        raise D128RecursivePCDRouteSelectorError(f"{field} must be repo-relative")
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)